### ISSUE TITLE
[TS-12] task focus

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskshell#11]
_John Florian on 2009-07-30T13:33:41Z says:_ 

 Although the .taskrc file provides an option for setting a default project, that seems best suited for situations where the user works predominately on a single project for long periods of time.  It would be very handy to be able to set a default project on the command line that stays in effect until overridden again.  For example:

<pre>
task focus foo    # more or less equivalent to rc.default.project=foo
<lots of task work on the project foo>
task focus bar    # more or less equivalent to rc.default.project=bar
<lots of task work on the project bar>
task focus        # more or less equivalent to rc.default.project=
<lots of task work on a variety of projects
</pre>

While focus is given to a particular project, reports, additions, etc. would be pegged to the project of focus.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2009-07-30T13:33:41Z
Modified: 2014-12-22T01:26:39Z
```

**Comment by taskwarrior** [comment]
_David Patrick on 2009-07-30T18:40:18Z says:_ 

 I like where that's going. 

I think an "expanded shell-mode":/issues/show/210 would also be brilliant for that. 
imagine;

   task shell add proj:camping +gear

Enter new task details [attrs] [tags]
Type "q" and hit <enter> to quit shell mode
task add project:camping +gear>

djp

**Comment by taskwarrior** [comment]
_Federico Hernandez on 2009-07-30T19:09:28Z says:_ 

 We should not limit focus to projects  but to all kind of "information" regarding task modifications like tags, etc. I would also say to make the "syntax" more similar to the usage elsewhere in the program:

task focus proj:bar   to focus on the project bar
task focus +tag       to focus on all tasks with the tag set

I guess focus could be seen as permanent "filters".

If we go further what about having a possibility to set default values for certain commands when needed. For example when I know that I will add 15 tasks to the proj:bar that should also have the tag +ernie and +bert.

Of course I could do the 15 task add followed by a task 1-15 proj:bar +ernie +bert

**Comment by taskwarrior** [comment]
_John Florian on 2009-07-30T19:29:49Z says:_ 

 Federico Hernandez wrote:
> We should not limit focus to projects  but to all kind of "information" regarding task modifications like tags, etc. I would also say to make the "syntax" more similar to the usage elsewhere in the program:
> 
> task focus proj:bar   to focus on the project bar
> task focus +tag       to focus on all tasks with the tag set
> 
> I guess focus could be seen as permanent "filters".
> 
> If we go further what about having a possibility to set default values for certain commands when needed. For example when I know that I will add 15 tasks to the proj:bar that should also have the tag +ernie and +bert.

I concur 100% Fredde.

> Of course I could do the 15 task add followed by a task 1-15 proj:bar +ernie +bert

That works fine for a batch of additions, but I'm often faced with mostly non-sequential IDs for any given project as the tasks get added over time.  I'll be working on something, but may have forgotten the ID, so I'll do a list for that one project.  I'll mark that task done and probably do another list, again just for that project, to find out what's next.  I'll begin working on the next ID and while looking at some code, realize I need to add a new task (to be done later) for the same project.  Adding that new task jostles something in my head so I'll need to annotate (same project) and so forth.

In short, almost every command I do will have pro:foo in there somewhere -- it's my primary method for aggregating my tasks.  I'd just like to not have to type that portion (likewise with tags) each and every time.  My "focus" will need to change periodically, but rarely on a per-task basis.

**Comment by taskwarrior** [comment]
_David Patrick on 2009-07-30T20:49:38Z says:_ 

 what about shell mode, as described above?
from the CLI I can't think of anything faster..
(you could use it with tags too)

**Comment by taskwarrior** [comment]
_John Florian on 2009-07-30T21:02:30Z says:_ 

 David Patrick wrote:
> what about shell mode, as described above?
> from the CLI I can't think of anything faster..
> (you could use it with tags too)

Sorry, I meant to respond to your shell idea -- I like that approach very much as well.  It would be best to provide both, of course.  While the shell is really nice for not having to escape special characters, I find myself not using it presently because it lacks command history and tab completion.  However, if it supported focus, history and tab completion become less important there.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-05-30T22:02:26Z says:_ 

 I think enhanced aliases could help here, but this needs more definition.
